File: 2020/04/04.py
import time

with open("input", 'r') as f:
  lines = f.readlines()

validPassports = 0
passportsProcessed = 0
first = True
needed_keys = ["byr", "iyr", "hgt", "ecl", "hcl", "pid", "eyr"]

def validate_byr(byr_v):
  i = int(byr_v)
  return (i >= 1920) and (i <= 2002)

def validate_iyr(iyr_v):
  i = int(iyr_v)
  return (i >= 2010) and (i <= 2020)

def validate_eyr(eyr_v):
  i = int(eyr_v)
  return (i >= 2020) and (i <= 2030)

def validate_hgt(hgt_v):
  try:
    val = int(hgt_v[:-2])
  except ValueError:
    return False
  unit = hgt_v[-2:]
  if unit == "cm":
    return (val >= 150) and (val <= 193)
  elif unit == "in":
    return (val >= 59) and (val <= 76)
  else:
    return False

def validate_ecl(ecl_v):
  return ecl_v in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']

def validate_hcl(hcl_v):
  return ((len(hcl_v) == 7)
    and (hcl_v[0] == '#')
    and (all([hcl_v[i] in "0123456789abcdef"
      for i in range(1,7)])))

def validate_pid(pid_v):
  return ((len(pid_v) == 9)
    and all([ch in "0123456789" for ch in pid_v]))

validator_d = {
  'byr': validate_byr,
  'eyr': validate_eyr,
  'iyr': validate_iyr,
  'hgt': validate_hgt,
  'hcl': validate_hcl,
  'ecl': validate_ecl,
  'pid': validate_pid,
}

def is_valid(pp):
  if len(pp) < 7:
    return False
  if all([key in pp.keys() for key in needed_keys]):
    return all([validator_d[key](passport[key]) for key in needed_keys])
  return False

t0 = time.perf_counter()

for line in lines:
  if line != '\n':
    if first:
      passport = {}
      first = False
    line = line[:-1]
    kvpairs = line.split(' ')
    for kv in kvpairs:
      k, v = kv.split(':')
      passport[k] = v
  else:
    if (is_valid(passport)):
      validPassports += 1
    first = True
    passportsProcessed += 1

if (is_valid(passport)):
  validPassports += 1

t1 = time.perf_counter()
elapsed = t1 - t0

print(f"Part one: {validPassports} of {passportsProcessed} ({1000 * elapsed:0.3f} usec)")

File: 2020/04/04.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <utility>
#include <cctype>
#include "timing.hpp"

std::vector<std::string> needed_keys = { "byr", "iyr", "hgt", "ecl", "hcl", "pid", "eyr" };
using Passport = std::unordered_map<std::string, std::string>;

bool byr(std::string byrv)
{
  auto i = stoi(byrv);
  return (i >= 1920 and i <= 2002);
}

bool iyr(std::string iyrv)
{
  auto i = stoi(iyrv);
  return (i >= 2010 and i <= 2020);
}

bool eyr(std::string eyrv)
{
  auto i = stoi(eyrv);
  return (i >= 2020 and i <= 2030);
}

bool hgt(std::string hgtv)
{
  auto len = hgtv.length();
  auto val = hgtv.substr(0, len - 2);
  if (not (all_of(begin(val), end(val), [](char ch){isdigit(ch);})))
  {
    return false;
  }
  else
  {
    auto val_i = stoi(val);
    auto unit = hgtv.substr(len - 2);
    if (unit == "cm") { return ((val_i >= 150) and (val_i <= 193)); }
    else if (unit == "in") { return (val_i >= 59) and (val_i <= 76); }
    else { return false; }
  }
}

bool ecl(std::string eclv)
{
  static std::vector<std::string> ecl_valid = {
    "amb", "blu", "brn", "gry", "grn", "hzl", "oth"
  };
  for (const auto& st : ecl_valid)
  {
    if (eclv == st) { return true; }
  }
  return false;
}

bool ishexdigit(char ch)
{
  switch (ch)
  {
    case '0':
    case '1':
    case '2':
    case '3':
    case '4':
    case '5':
    case '6':
    case '7':
    case '8':
    case '9':
    case 'a':
    case 'b':
    case 'c':
    case 'd':
    case 'e':
    case 'f': return true;
    default : return false;
  }
}

bool hcl(std::string hclv)
{
  auto it = begin(hclv);
  return ((hclv.length() == 7) and (hclv[0] == '#')
    and all_of(++it, end(hclv), ishexdigit));
}

bool pid(std::string pidv)
{
  return (pidv.length() == 9) and all_of(begin(pidv), end(pidv), isdigit);
}

std::vector<bool (*)(std::string)> validators_v = {
  &byr, &iyr, &hgt, &ecl, &hcl, &pid, &eyr // same order as needed_keys
};

bool is_valid(Passport pp)
{
  if (pp.size() < 7) { return false; }
  for (size_t i = 0; i < validators_v.size(); ++i)
  {
    if (!(validators_v[i](needed_keys[i]))) { return false; }
  }
  return true;
}

int main()
{
  std::ifstream ifs;
  ifs.open("input");
  bool first = true;
  Passport pp;
  for(std::string line; getline(ifs, line); )
  {
    if (line != "")
    {
      if (first)
      {
        pp = Passport();
        first = false;
      }
      std::ostringstream iss;
      iss << line;
    }
  }
}
